### DISCUSSION TITLE
Can the calculated results in Postprocessors modular be used for calculations in materials modular?

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#19650] [Q&A General] [answered]
Dear MOOSE experts,

For the same purpose in the problem #19648, here is a plan B that can be used.  So, I would like to ask whether the average value calculated in the Postprocessors module can be used to calculate a certain `MaterialProperty`  variable in the materials module.

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by GiudGiud** [comment, marked as answer]
Hello

yes you can use postprocessors when computing material properties. 
just save a reference to the postprocessor value with `getPostprocessorValueByName(pp_name)` with pp_name obtained from the input parameters. 

However, please be mindful that if you are using AD, postprocessors will not transfer the derivative information to your material property. 

Guillaume

**Reply by PengWei97** [reply]
Ok, Thank you for your very kind reply.

I got it. Is it possible to couple `variable` (such as, gr#, the order parameter in phase field) in `ElementAverageMaterialProperty`, and then define the average value as,
> sum_value += material_property[_qp] * gr0 *gr0;
> area += element_area * gr0 *gr0;
> average_value = sum_value /area ;

This is what I want to try next, if it can be coupled. What I really need to be weighted and average is only a material parameter in each grain, #19648.

**Reply by hugary1995** [reply]
Yes, you can couple variables in objects derived from `ElementUserobjects`. But notice that you'll need to sum the material property and area separately in the `execute()` method, and compute the average in the `getValue()` method.

**Reply by PengWei97** [reply]
Ok, I know. I think that if the operation of performing the average value is executed in the `Userobjects` block or an `postprocessing` block, it does not seem to be a wise choice to execute it in the material module.

 Best Wishes
Wei
